src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php

```diff
@@ -319,8 +320,13 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
                         $params = array_merge($params, $data[$paramName]);
                     }
                 } elseif ($allowed && !$ignored && (isset($data[$key]) || array_key_exists($key, $data))) {
-                    $params[] = $data[$key];
-                    // don't run set for a parameter passed to the constructor
+                    $parameterData = $data[$key];
+                    if (null !== $constructorParameter->getClass()) {
+                        $parameterData = $this->serializer->denormalize($parameterData, $constructorParameter->getClass()->getName(), null, $context);
```

Comment:
Also calling `deserialize()` will fail if you are using a serializer without any encoder, which might be the case when you are dealing with arrays and php objects, i.e. are not taking any JSON/XML input. Not likely in a regular app, but can happen in libraries.